cmd ref: get: document --show-url flag (#936)
iterative/dvc.org: public/static/docs/command-reference/get.md

  an existing directory is specified, then the output will be placed inside of
  it.

> - `--rev` - `url` is expected to contain a Git repository for this option to

Reviewer: I would say in this case we don't need to specify these "pre-req" at all - neither before or after :) May be something like "Obviously, a Git repository must be located at provide location (`url` arg)." at the end.

Reviewer 2: Yeah I guess it's obvious. No need to add the "obviously" note though, doesn't sound friendly. Removed this prereq sentence.

Reviewer 2: Applied the same for `dvc import`.

  have an effect. Specific
  [Git revision](https://git-scm.com/book/en/v2/Git-Internals-Git-References)
  (such as a branch name, a tag, or a commit hash) of the repository to download
  the file or directory from. The tip of the default branch is used by default
  when this option is not specified.

> - `--show-url` - `path` is expected to represent a cached, DVC-tracked file for

Reviewer: same here, still would prefer to see the essence of it. After that something like: Limitations: it does not print anything (or throws an exception?) for data files or directories that are not DVC-tracked (usually, saved into Git directly instead). See example ... link"

Reviewer 2: OK, moved the prereq to last in cab347ca.

Reviewer 2: And made standard use of this order in other commands.

Reviewer 2: > Limitations: it does not print anything...

p.s. I think it's best to just say what the prereq is and not what the command will do if the prereq is not met. Easier text to maintain.

  the file or directory from. The tip of the default branch is used by default
  when this option is not specified.

- `--show-url` - `url` is expected to represent a DVC project for this option to
  have an effect (as opposed to a Git-only, non-DVC repository). When
  `--show-url` is used, instead of downloading the file or directory, this

>   command just prints the storage location (URL) of the target data.

Reviewer: I still think we are overcomplicating this. The fact that it does not handle Git repos is only somewhat important and at least should go after the essence of what this command does. Also, probably it's not about DVC repo vs Git repo - it's about Git-tracked files vs DVC-tracked files. Am I correct here, @skshetry ?

Also, consider including a link the example below.

Author: No, it's about DVC repo vs Git repo. As an example, take this *git* repo. How can you get an url for this `public/static/docs/command-reference/get.md`? Git does not work with files, so there's no URL/path to generate for (unless we do somekind of *magic*).

Reviewer: okay, to clarify it's not about Git-only vs DVC. Any Git+DVC has files that are not under DVC control and probably we don't generate URL for them, right?

Reviewer: I'm not even sure about DVC-controlled non-cached files (`dvc run -O something`), probably we don't even handle that with this new option (even though we can download it)

Author: Yes, that should not also work, as it checks for all DVC controlled outputs.

Author: So, it's also about Git vs DVC tracked files. Sorry about that.

Reviewer 2: > The fact that it does not handle Git repos is only somewhat important and at least should go after the essence of what this command does.

@shcheklein You may be right but about this we're already using this style in most other options when there are pre-requisites. Putting the pre-req's as the first sentence. (See `--rev` just above this.) That's not a great reason to keep this but I'm also not convinced that leaving the pre-req's for last is best. Maybe it is better to leave them on top? they are "pre" requisites after all 🙂

All that said I did shorten that first sentence now, and also the main explanation following it.

> to clarify it's not about Git-only vs DVC. Any Git+DVC has files that are not under DVC control

True, I updated that as well. So the pre-req is not about `url` then, but about `path`: it should be a DVC-tracked file/dir.

See ab40ed7f.

Reviewer 2: > I'm not even sure about DVC-controlled non-cached files (dvc run -O something), probably we don't even handle that with this new option (even though we can download it)

This sounds like something I should note here and in `dvc.api.get_url` as well?

Reviewer 2: > I'm not even sure about DVC-controlled non-cached files (dvc run -O something), probably we don't even handle that with this new option (even though we can download it)

Actually, both this option and `get_url()` should support creating a URL for these. They don't atually check for the existence of the file in remote cache. They just interpret the project configured remote + the checksum in DVC-file from what I understand.

  the file or directory from. The tip of the default branch is used by default
  when this option is not specified.

> - `--show-url` - Returns path or url to the location in remote for given path,

Reviewer: `Returns` -> `returns`

Reviewer: why is it `path or ..`?

I would simplify significantly the whole piece.

"instead of downloading the file or directory print its actual location (URL) it would be downloaded from, e.g. s3://remote/storage/0c/121243234"

Reviewer 2: Agree with Ivan. This could start:
```suggestion
- `--show-url` - returns URL to the remote location of the target data
```
but the rest of the paragraph can be simplified as well...

Author: @jorgeorpinel, is the distinction not important, that, it can also be a path, not just the URL?

Reviewer 2: It's already confusing enough that the command takes both `url` and `path`, `url` accepting URLs and paths (explained in description), and then on top of that saying this option returns a "path or URL".
But OK I guess you're right, maybe check out the paragraph about the `url` command param and use similar language to explain this returned URL could be a local file system path (for local remotes).
OR don't use either path nor URL and just say something like "storage location" with URL corresponding to the type of remote, linking to remote cmd ref, where the "local" case is explained.

Reviewer: @jorgeorpinel @skshetry what about the way I suggested it? One sentence, using terms users would understood? `instead of downloading the file or directory print its actual location (URL) it would be downloaded from, e.g. s3://remote/storage/0c/121243234`? One simple sentence + even example to put more context.

Reviewer 2: Yeah I think you got it right Ivan. I added to the core repo in https://github.com/iterative/dvc/pull/3220/commits/ee0193ccd36fdc067ace131958f7960819202643 and will update here as well.

Reviewer 2: Updated in 072defa3.

  from where the **file** will be downloaded. `path` is used to specify the location
  of the target artifact within the source project in `url`, relative to the project's
  root. `url` can be an address to a DVC project or a path to a local DVC project.
  When this option is specified, it will not download the files.

>   Git-only repository is not supported.

Author: Probably, non-DVC repository would have been better?

Reviewer: let's not explain the positional args (`path` and `url` here) - they have the same meaning and are probably already described, right?

Reviewer: and/or even put a link to the example below ...

Reviewer 2: Yep, `url` and `path` are explained in the description. No need to do it again here, you can simply reference them. You can even use ``(`url`/`path`)`` as in the introductory paragraph.

Reviewer 2: Oh and
> Probably, "non-DVC repository"...

Yes, I like that term better. But typically we begin with applicability/restriction sentences in option paragraphs, for example see the language used here:
![image](https://user-images.githubusercontent.com/1477535/72761955-6517fb80-3ba3-11ea-9032-aaba43fda741.png)

Thanks

Reviewer 2: Updated in 072defa3.

	model.bigrams.pkl
	model.monograms.pkl
```

> ### Example: Getting a url for a file in a DVC repository

Reviewer 2: Also done in 072defa3.

> > Note that `dvc get` can be used from anywhere in the file system, as long as

Reviewer: have some doubts ... does not feel we need to repeat it here ... why did you decide to put it?

Reviewer 2: You mean the note about "can be used from anywhere in the file system" right Ivan? Not the whole example

Reviewer 2: Yaah, agree about the note. No need 🙂

Reviewer 2: Removed in 072defa3.

> DVC is [installed](/doc/install).

We can use `dvc get --show-url` to get the actual location of the resulting model file from our
[get started example repo](https://github.com/iterative/example-get-started), a

> <abbr>DVC project</abbr> hosted on GitHub:

Reviewer 2: Suggestion, not necessarily to use exactly:

``We can use `dvc get --show-url` to get the actual location where the final model file from our [get started example repo](https://github.com/iterative/example-get-started) is stored. Note that `remote.dvc.org/get-started` is an HTTP [DVC remote](/doc/command-reference/remote), whereas `662eb7f64216d9c2c1088d0a5e2c6951` is the file's checksum.``

Reviewer 2: Updated in 663f589a.
